service/src/builder.rs

```diff
@@ -291,22 +199,139 @@ fn ensure_binary(package: &Package) -> anyhow::Result<()> {
 
 /// Make sure "cdylib" is set for shuttle-next projects, else set it if possible.
 fn ensure_cdylib(package: &Package) -> anyhow::Result<()> {
-    if package.targets().iter().any(|target| target.is_lib()) {
+    if package.targets.iter().any(is_cdylib) {
         Ok(())
     } else {
         bail!("Your Shuttle next project must be a library. Please add `[lib]` to your Cargo.toml file.")
     }
 }
 
-/// Ensure `panic = "abort"` is not set:
-fn check_no_panic(ws: &Workspace) -> anyhow::Result<()> {
-    if let Some(profiles) = ws.profiles() {
-        for profile in profiles.get_all().values() {
-            if profile.panic.as_deref() == Some("abort") {
-                return Err(anyhow!("Your Shuttle project cannot have panics that abort. Please ensure your Cargo.toml does not contain `panic = \"abort\"` for any profiles."));
+fn is_cdylib(target: &Target) -> bool {
+    target.kind.iter().any(|kind| kind == "cdylib")
+}
+
+async fn compile(
+    packages: Vec<&Package>,
+    release_mode: bool,
+    wasm: bool,
+    project_path: PathBuf,
+    deployment: bool,
+    tx: Sender<Message>,
+) -> anyhow::Result<Vec<BuiltService>> {
+    let manifest_path = project_path.join("Cargo.toml");
+
+    let mut cargo = tokio::process::Command::new("cargo");
+
+    let (mut reader, writer) = os_pipe::pipe()?;
+    let writer_clone = writer.try_clone()?;
+    cargo.stdout(writer);
+    cargo.stderr(writer_clone);
+
+    cargo.arg("build").arg("--manifest-path").arg(manifest_path);
+
+    if deployment {
+        cargo.arg("-j").arg(4.to_string());
+    }
+
+    for package in packages.clone() {
+        cargo.arg("--package").arg(package.name.clone());
+    }
+
+    let mut profile = "debug";
+
+    if release_mode {
+        profile = "release";
+        cargo.arg("--profile").arg("release");
+    } else {
+        cargo.arg("--profile").arg("dev");
+    }
+
+    if wasm {
+        cargo.arg("--target").arg("wasm32-wasi");
+    }
+
+    let mut handle = cargo.spawn()?;
+
+    tokio::task::spawn_blocking(move || {
+        let mut paused = false;
+        let mut buf = [0; 30];
+        let mut read: usize = 30;
+        loop {
+            if read == 0_usize {
+                if paused {
+                    println!("broken");
+                    break;
+                }
+                std::thread::sleep(std::time::Duration::from_millis(10000));
+                paused = true;
             }
+            read = reader.read(&mut buf).unwrap();
+            if let Err(error) = tx.send(Message::TextLine(
+                String::from_utf8(buf.clone().to_vec()).unwrap(),
+            )) {
+                error!("failed to send cargo message on channel: {error}");
+            };
+        }
+    });
```

Comment:
Can we rather read lines here? This should work, let me know if you tried this already and you found some issues with it!
```suggestion
    tokio::task::spawn_blocking(move || {
        let reader = std::io::BufReader::new(reader);
        for line in reader.lines() {
            if let Ok(line) = line {
                if let Err(error) = tx.send(Message::TextLine(line)) {
                    error!("failed to send cargo message on channel: {error}");
                };
            } else {
                // TODO: handle error
                println!("err reading line");
            }
        }
    });
```

Reply:
Yeah we could, my weird solution is the result of doing this at midnight.